### PR TITLE
[FLINK-7199] [gelly] Graph simplification does not set parallelism

### PR DESCRIPTION
**Pull request by greghogan**
The Simplify parameter should accept and set the parallelism when calling the Simplify algorithms.
    
The LocalClusteringCoefficient "count triangles" reduce now uses the assigned ("little") parallelism as this computation is proportional to the number of vertices (the combine computation is proportional to the potentially much larger number of triangles).

The ignored CombineHint on the HITS all-reduces have been removed.

### CODE REVIEW

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/input/InputBase.java [hunk 1]
```diff
@@ -32,6 +35,9 @@
 extends ParameterizedBase
 implements Input<K, VV, EV> {
 
+	protected LongParameter parallelism = new LongParameter(this, "__parallelism")
+		.setDefaultValue(PARALLELISM_DEFAULT);
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/input/InputBase.java]: I know, it's been that way before, but what a `LongParameter` although only the `intValue()` of it is actually used?

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/linkanalysis/HITS.java [hunk 2]
```diff
@@ -168,7 +168,6 @@ protected void mergeConfiguration(GraphAlgorithmWrappingBase other) {
 				.setParallelism(parallelism)
 				.name("Square")
 			.reduce(new Sum())
-			.setCombineHint(CombineHint.HASH)
```
Comment (NicoK) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/linkanalysis/HITS.java]: this changes doesn't look related to the intention of the commit, or did I miss something?

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/linkanalysis/HITS.java [hunk 3]
```diff
@@ -192,7 +191,6 @@ protected void mergeConfiguration(GraphAlgorithmWrappingBase other) {
...
```
Comment (NicoK) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/linkanalysis/HITS.java]: same here

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/Simplify.java [hunk 4]
```diff
@@ -109,21 +109,23 @@ public Ordering getValue() {
...
-	public <T extends Comparable<T>> Graph<T, NullValue, NullValue> simplify(Graph<T, NullValue, NullValue> graph)
+	public <T extends Comparable<T>> Graph<T, NullValue, NullValue> simplify(Graph<T, NullValue, NullValue> graph, int parallelism)
 			throws Exception {
-
 		switch (value) {
 			case DIRECTED:
 				graph = graph
-					.run(new org.apache.flink.graph.asm.simple.directed.Simplify<T, NullValue, NullValue>());
+					.run(new org.apache.flink.graph.asm.simple.directed.Simplify<T, NullValue, NullValue>()
+						.setParallelism(parallelism));
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/Simplify.java]: the indention of the second call should be the same as the first one

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/Simplify.java [hunk 5]
```diff
@@ -109,21 +109,23 @@ public Ordering getValue() {
...
 				break;
 			case UNDIRECTED:
 				graph = graph
-					.run(new org.apache.flink.graph.asm.simple.undirected.Simplify<T, NullValue, NullValue>(false));
+					.run(new org.apache.flink.graph.asm.simple.undirected.Simplify<T, NullValue, NullValue>(false)
+						.setParallelism(parallelism));
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/Simplify.java]: same here...

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/Simplify.java [hunk 6]
```diff
@@ -109,21 +109,23 @@ public Ordering getValue() {
...
 				break;
 			case UNDIRECTED_CLIP_AND_FLIP:
 				graph = graph
-					.run(new org.apache.flink.graph.asm.simple.undirected.Simplify<T, NullValue, NullValue>(true));
+					.run(new org.apache.flink.graph.asm.simple.undirected.Simplify<T, NullValue, NullValue>(true)
+						.setParallelism(parallelism));
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/Simplify.java]: ...and here

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/input/InputBase.java (hunk 1 above)
Comment (greghogan) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/input/InputBase.java]: We didn't write an `IntParameter`. These are only used on the client when parsing command-line parameters.

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/linkanalysis/HITS.java (hunk 2 above)
Comment (greghogan) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/linkanalysis/HITS.java]: I had updated the commit message which was not reflected in the pull request message. I've updated the latter.

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/Simplify.java (hunk 4 above)
Comment (greghogan) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/Simplify.java]: This is consistent throughout Gelly as we often chain operators.

Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/Simplify.java]: it's different in the changes to `LocalClusteringCoefficient.java` though - was just wondering

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/input/InputBase.java (hunk 1 above)
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/input/InputBase.java]: thanks for the explanation

Comment (greghogan) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/input/InputBase.java]: This should have the `min` and `max` values set.

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/linkanalysis/HITS.java (hunk 2 above)
Comment (NicoK) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/linkanalysis/HITS.java]: would have expected a separate commit but ok - I can't say much about the change though since I don't know flink's graph API too much...

**Comment by greghogan** [comment]
@NicoK thanks for the review! I'll work on refactoring the `Runner` and adding checks for checking the parallelism. I think we're okay if we whitelist the "large" operators by name.

**Comment by greghogan** [comment]
@NicoK, parallelism tests have been added. There is a small refactor for `Runner` which needs more extensive rework elsewhere.

File: flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java [hunk 7]
```diff
@@ -56,214 +56,304 @@ public void testLongDescription() throws Exception {
 			ProgramParametrizationException.class);
 	}
 
+	// CirculantGraph
+
+	private String[] getCirculantGraphParamters(String output) {
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java]: typo, should be `getCirculantGraphParameters`

File: flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java [hunk 8]
```diff
@@ -56,214 +56,304 @@ public void testLongDescription() throws Exception {
...
+	}
+
+	@Test
+	public void testParallelismWithCirculantGraph() throws Exception {
+		TestUtils.verifyParallelism(getCirculantGraphParamters("print"));
+	}
+
+	// CompleteGraph
+
+	private String[] getCompleteGraphParamters(String output) {
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java]: typo, should be: `getCompleteGraphParameters`

File: flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java [hunk 9]
```diff
@@ -56,214 +56,304 @@ public void testLongDescription() throws Exception {
...
+	}
+
+	@Test
+	public void testParallelismWithCycleGraph() throws Exception {
+		TestUtils.verifyParallelism(getCycleGraphParamters("print"));
+	}
+
+	// EchoGraph
+
+	private String[] getEchoGraphParamters(String output) {
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java]: same here: `getEchoGraphParameters`

File: flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java [hunk 10]
```diff
@@ -56,214 +56,304 @@ public void testLongDescription() throws Exception {
...
+	}
+
+	@Test
+	public void testParallelismWithEchoGraph() throws Exception {
+		TestUtils.verifyParallelism(getEchoGraphParamters("print"));
+	}
+
+	// EmptyGraph
+
+	private String[] getEmptyGraphParamters(String output) {
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java]: and here: `getEmptyGraphParameters`

File: flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java [hunk 11]
```diff
@@ -56,214 +56,304 @@ public void testLongDescription() throws Exception {
...
+	}
+
+	@Test
+	public void testParallelismWithEmptyGraph() throws Exception {
+		TestUtils.verifyParallelism(getEmptyGraphParamters("print"));
+	}
+
+	// GridGraph
+
+	private String[] getGridGraphParamters(String output) {
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java]: one more: `getGridGraphParameters`

File: flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java [hunk 12]
```diff
@@ -56,214 +56,304 @@ public void testLongDescription() throws Exception {
...
+	}
+
+	@Test
+	public void testParallelismWithGridGraph() throws Exception {
+		TestUtils.verifyParallelism(getGridGraphParamters("print"));
+	}
+
+	// HypercubeGraph
+
+	private String[] getHypercubeGraphParamters(String output) {
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java]: that typo seems consistent... `getHypercubeGraphParameters`

File: flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java [hunk 13]
```diff
@@ -56,214 +56,304 @@ public void testLongDescription() throws Exception {
...
+	}
+
+	@Test
+	public void testParallelismWithHypercubeGraph() throws Exception {
+		TestUtils.verifyParallelism(getHypercubeGraphParamters("print"));
+	}
+
+	// PathGraph
+
+	private String[] getPathGraphParamters(String output) {
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java]: `getPathGraphParameters`

File: flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java [hunk 14]
```diff
@@ -56,214 +56,304 @@ public void testLongDescription() throws Exception {
...
+	}
+
+	@Test
+	public void testParallelismWithPathGraph() throws Exception {
+		TestUtils.verifyParallelism(getPathGraphParamters("print"));
+	}
+
+	// RMatGraph
+
+	private String[] getRMatGraphParamters(String output, String simplify) {
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java]: `getRMatGraphParameters`

File: flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java [hunk 15]
```diff
@@ -56,214 +56,304 @@ public void testLongDescription() throws Exception {
...
+	}
+
+	@Test
+	public void testParallelismWithUndirectedRMatGraph() throws Exception {
+		TestUtils.verifyParallelism(getRMatGraphParamters("print", "undirected"));
+	}
+
+	// SingletonEdgeGraph
+
+	private String[] getSingletonEdgeGraphParamters(String output) {
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java]: `getSingletonEdgeGraphParameters`

File: flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java [hunk 16]
```diff
@@ -56,214 +56,304 @@ public void testLongDescription() throws Exception {
...
+	}
+
+	@Test
+	public void testParallelismWithSingletonEdgeGraph() throws Exception {
+		TestUtils.verifyParallelism(getSingletonEdgeGraphParamters("print"));
+	}
+
+	// StarGraph
+
+	private String[] getStarGraphParamters(String output) {
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/test/java/org/apache/flink/graph/drivers/EdgeListITCase.java]: `getStarGraphParameters`

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/LongParameter.java [hunk 17]
```diff
@@ -52,16 +52,6 @@ public LongParameter(ParameterizedBase owner, String name) {
...
-		if (hasMinimumValue) {
-			Util.checkParameter(defaultValue >= minimumValue,
-				"Default value (" + defaultValue + ") must be greater than or equal to minimum (" + minimumValue + ")");
-		}
-
-		if (hasMaximumValue) {
-			Util.checkParameter(defaultValue <= maximumValue,
-				"Default value (" + defaultValue + ") must be less than or equal to maximum (" + maximumValue + ")");
-		}
-
```
Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/LongParameter.java]: I'm no expert here, but can you elaborate on why you removed these checks for the default value being in line with the min and max?

I can only guess that this must be that way in case where the default parallelism is not within the bounds of the configured one. I'm wondering whether the `LongParameter` is used somewhere else were these bounds are expected to hold though.

Comment (greghogan) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/LongParameter.java]: Right, `ExecutionConfig.DEFAULT_PARALLELISM` is currently `-1`. We're still checking the user configuration within the min and max bounds, it's only when no value is given that the default is used and now permitted to lie outside of the bounds.

Comment (NicoK) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/drivers/parameter/LongParameter.java]: ok, should we make this explicit in the `LongParamater` class' javadoc? For example:
```
/**
 * A {@link Parameter} storing a {@link Long} within <tt>min</tt> and <tt>max</tt> bounds (inclusive).
 *
 * <p>Note that the default value may be outside of these bounds.
 */
```

**Comment by greghogan** [comment]
@NicoK I have fixed the noted issues and will merge.
